### PR TITLE
Search: Search Functionality added on index and Docs page.

### PR DESCRIPTION
**Pull request by Bazif-Khan**
## Checklist

- [x] commit message follows [commit guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md#commit-message-guidelines).
- [x] website builds and runs in production - for information on how to test locally go [here](https://github.com/appsody/website/blob/master/DEVELOPMENT.md).

## Related Issues

#issue-number

### CODE REVIEW

**Comment by tomleah** [comment]
Can we use camelcase for file names. `Search.js` -> `search.js`

**Comment by Bazif-Khan** [comment]
> Can we use camelcase for file names. `Search.js` -> `search.js`

I have changed it.

File: package.json [hunk 1]
```diff
@@ -5,6 +5,7 @@
   "version": "0.1.0",
   "dependencies": {
     "@fortawesome/fontawesome-free": "^5.9.0",
+    "axios": "^0.19.0",
```
Comment (tomleah) [inline package.json]: This dependency is not used

File: gatsby-config.js [hunk 2]
```diff
@@ -19,6 +19,13 @@ module.exports = {
         anonymize: true,
       }
     },
+    {
+      resolve: `gatsby-source-filesystem`,
+      options: {
+        name: `sidebaryaml`,
+        path: `${__dirname}/content/docs`,
```
Comment (tomleah) [inline gatsby-config.js]: Do we need this still?

File: package.json (hunk 1 above)
Comment (tomleah) [reply package.json]: Run `npm prune` after deleting the dependency to clean up `package-lock.json`

File: src/components/Search.js [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+    this.setState({ search: dataToSearch, isLoading: false })
+  }
+
+  /**
+   * handles the input change and perform a search with js-search
+   * in which the results will be added to the state
+   */
+  searchData = e => {
+    const { search } = this.state
```
Comment (tomleah) [inline src/components/Search.js]: `search` is not used here

File: src/components/Search.js [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    const queryResult = search.search(e.target.value)
+
+    this.setState({ searchQuery: e.target.value, searchResults: queryResult })
+  }
+  handleSubmit = e => {
+    e.preventDefault()
+  }
+
+  closeDropdown() {
+    this.setState(state => ({
```
Comment (tomleah) [inline src/components/Search.js]: `state` is not used here

File: src/components/Search.js [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+      isOpen: false
+    }));
+  }
+
+  openDropdown() {
+    this.setState(state => ({
```
Comment (tomleah) [inline src/components/Search.js]: `state` is not used here

File: src/components/Search.js [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+      isOpen: true
+    }));
+  }
+
+  render() {
+    const { titleList, searchResults, searchQuery } = this.state
```
Comment (tomleah) [inline src/components/Search.js]: `titleList` is not used here

File: src/components/Search.js [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+          </form>
+          <div id="search-dropdown" className="dropdown-content ">
+            <table>
+            {queryResults.map(item => {
+                return (
+                <tr key={`row_${item.title}`}>
+                  <td>
+                    <Link onClick={this.closeDropdown} to={item.frontmatter.path}>{item.frontmatter.title}</Link>
+                  </td>
+
```
Comment (tomleah) [inline src/components/Search.js]: nit: remove new line

File: src/components/Search.js [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+          <form className="form-inline" onSubmit={this.handleSubmit}>
+              <input id="search-input" onClick={this.openDropdown} className="form-control form-control-sm mt-2 mr-2" type="text" placeholder="Search"
+                aria-label="Search"
+                id="Search"
+                value={searchQuery}
+                onChange={this.searchData}
+              />
+              <i class="fa fa-search ml-2 mt-1 mr-3"></i>
+          </form>
+          <div id="search-dropdown" className="dropdown-content ">
```
Comment (tomleah) [inline src/components/Search.js]: nit: remove space in `className`

File: src/components/Search.js [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+          </form>
+          <div id="search-dropdown" className="dropdown-content hide">
+            <table>
+            {queryResults.map(item => {
+                return (
+                <tr key={`row_${item.title}`}>
+                  <td>
+                    <Link onClick={this.closeDropdown} to={item.frontmatter.path}>{item.frontmatter.title}</Link>
+                  </td>
+
```
Comment (tomleah) [inline src/components/Search.js]: nit: remove new line

File: src/styles/main.css [hunk 10]
```diff
@@ -361,4 +361,52 @@ footer a:hover {
...
+  overflow: auto;
+  border: 1px solid #ddd;
+  z-index: 1;
+  max-width: 13.1em;
+  max-height: 60vh;
+}
+
+.dropdown-content a {
+  color: black;
+  padding: 12px 16px;
```
Comment (tomleah) [inline src/styles/main.css]: Could we use `em` here?

File: package.json (hunk 1 above)
Comment (Bazif-Khan) [reply package.json]: Removed

File: gatsby-config.js (hunk 2 above)
Comment (Bazif-Khan) [reply gatsby-config.js]: No, I just forgot to remove this. I have removed this now.

File: src/components/Search.js (hunk 3 above)
Comment (Bazif-Khan) [reply src/components/Search.js]: search is used in line number 66.

File: src/components/Search.js (hunk 6 above)
Comment (Bazif-Khan) [reply src/components/Search.js]: removed

File: src/components/Search.js (hunk 7 above)
Comment (Bazif-Khan) [reply src/components/Search.js]: removed

File: src/components/Search.js (hunk 8 above)
Comment (Bazif-Khan) [reply src/components/Search.js]: removed

File: src/styles/main.css (hunk 10 above)
Comment (Bazif-Khan) [reply src/styles/main.css]: changed to em.

File: src/components/Search.js (hunk 9 above)
Comment (Bazif-Khan) [reply src/components/Search.js]: removed

File: src/components/Search.js (hunk 4 above)
Comment (Bazif-Khan) [reply src/components/Search.js]: I have removed the state.

File: src/components/Search.js (hunk 5 above)
Comment (Bazif-Khan) [reply src/components/Search.js]: I have removed the state.

File: src/components/Search.js (hunk 3 above)
Comment (tomleah) [reply src/components/Search.js]: Sorry I didn't see that 🙂

File: src/components/search.js [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+      return (
+        <div className="dropdown">
+          <form className="form-inline" onSubmit={this.handleSubmit}>
+              <input id="search-input" onClick={this.openDropdown} className="form-control form-control-sm mt-2 mr-2" type="text" placeholder="Search"
+                aria-label="Search"
+                id="Search"
+                value={searchQuery}
+                onChange={this.searchData}
+              />
+              <i class="fa fa-search ml-2 mt-1 mr-3"></i>
```
Comment (tomleah) [inline src/components/search.js]: `class` -> `className`

File: src/components/search.js [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
```
Comment (tomleah) [inline src/components/search.js]: `class` -> `className`

**Comment by tomleah** [comment]
I'm off on vacation till 19th August now.
Maintainers, here is where im currently at:
- I expect 1 error on the client (see `master` branch)
- I expect 0 warnings and errors on the server console when running `gatsby develop`

There is 1 warning in the console that is `Do not mutate state directly. Use setState()`.  The PR doesn't use `setState()` because it doesn't work within `componentDidmount()`. I would like to investigate this further to remove this warning. You can easily remove the warning but the functionality then breaks so be careful. 
You could try...wrapping `setState()` in a async function like `setTimeout`
```
setTimeout(() => {
     this.setState({pageList: this.props.items});
}, 0);
```
I haven't tried the above, and it seems super hacky but arguabily better than setting state incorrectly. 

Happy to merge with `Do not mutate state directly. Use setState()` if the above has been tried and there is some investigation, along with an issue raised of course 🙂

**Note:**
- client console is accessed in the browser's development tools
- server console is accessed by running `gatsby develop`
- If in doubt, compare errors to master

File: src/components/search.js (hunk 11 above)
Comment (Bazif-Khan) [reply src/components/search.js]: Changed

File: src/components/search.js (hunk 12 above)
Comment (Bazif-Khan) [reply src/components/search.js]: Changed

**Comment by Bazif-Khan** [comment]
All errors are fixed.

**Pull request closed without merging**